hass_nabucasa/iot_base.py

```diff
@@ -0,0 +1,240 @@
+
+    @property
+    def connected(self):
+        """Return if we're currently connected."""
+        return self.state == STATE_CONNECTED
+
+    async def async_send_json_message(self, message):
+        """Send a message.
+
+        Raises NotConnected if client not connected.
+        """
+        if self.state != STATE_CONNECTED:
+            raise NotConnected
+
+        if self._logger.isEnabledFor(logging.DEBUG):
+            self._logger.debug("Publishing message:\n%s\n", pprint.pformat(message))
+
+        await self.client.send_json(message)
+
+    async def connect(self):
+        """Connect to the IoT broker."""
+        if self.state != STATE_DISCONNECTED:
+            raise RuntimeError("Connect called while not disconnected")
+
+        self.close_requested = False
+        self.state = STATE_CONNECTING
+        self.tries = 0
+
+        while True:
+            try:
+                await self._handle_connection()
+            except Exception:  # pylint: disable=broad-except
+                # Safety net. This should never hit.
+                # Still adding it here to make sure we can always reconnect
+                self._logger.exception("Unexpected error")
+
+            if self.state == STATE_CONNECTED and self._on_disconnect:
+                try:
+                    await asyncio.wait([cb() for cb in self._on_disconnect])
+                except Exception:  # pylint: disable=broad-except
+                    # Safety net. This should never hit.
+                    # Still adding it here to make sure we don't break the flow
+                    self._logger.exception(
+                        "Unexpected error in on_disconnect callbacks"
+                    )
+
+            if self.close_requested:
+                break
+
+            self.state = STATE_CONNECTING
+            self.tries += 1
+
+            try:
+                await self._wait_retry()
+            except asyncio.CancelledError:
+                # Happens if disconnect called
+                break
+
+        self.state = STATE_DISCONNECTED
+
+    async def _wait_retry(self):
+        """Wait until it's time till the next retry."""
+        # Sleep 2^tries + 0…tries*3 seconds between retries
+        self.retry_task = self.cloud.run_task(
+            asyncio.sleep(2 ** min(9, self.tries) + random.randint(0, self.tries * 3))
+        )
+        await self.retry_task
+        self.retry_task = None
+
+    async def _handle_connection(self):
+        """Connect to the IoT broker."""
+        try:
+            await self.cloud.run_executor(self.cloud.auth.check_token)
+        except Unauthenticated as err:
+            self._logger.error("Unable to refresh token: %s", err)
+
+            self.cloud.client.user_message(
+                "cloud_subscription_expired", "Home Assistant Cloud", MESSAGE_AUTH_FAIL
+            )
+
+            # Don't await it because it will cancel this task
+            self.cloud.run_task(self.cloud.logout())
```

Comment:
We should not run that code on every service. Maybe we should offload the auth token handling/update/subscription and check those things once on cloud/init.

Reply:
Well this is when we refresh the token . I guess we can put it inside the refresh token code.